src/test/app/Offer_test.cpp

```diff
@@ -795,22 +850,831 @@ class Offer_test : public beast::unit_test::suite
+            jrr[jss::node][sfBalance.fieldName] == XRP (250).value ().getText ()
+        );
+
+        jrr = ledgerEntryState (env, bob, gw, "USD");
+        BEAST_EXPECT(jrr[jss::node][sfBalance.fieldName][jss::value] == "-400");
+    }
+
+    void
+    testcaseSellFlagExceedLimit()
+    {
+        testcase ("Offer tfSell: 2x Sell Exceed Limit");
+
+        using namespace jtx;
+
+        Env env {*this};
+        auto const gw = Account {"gateway"};
+        auto const alice = Account {"alice"};
+        auto const bob = Account {"bob"};
+        auto const USD = gw["USD"];
+
+        auto const starting_xrp = XRP (100) +
+            env.current ()->fees ().accountReserve (1) +
+            env.current ()->fees ().base * 2;
+
+        env.fund (starting_xrp, gw, alice, bob);
+
+        env (trust (alice, USD (150)));
+        env (trust (bob, USD (1000)));
+
+        env (pay (gw, bob, bob["USD"] (500)));
+
+        env (offer (bob, XRP (100), USD (200)));
+        // Alice has 350 fees - a reserve of 50 = 250 reserve = 100 available.
+        // Ask for more than available to prove reserve works.
+        // Taker pays 100 USD for 100 XRP.
+        // Selling XRP.
+        // Will sell all 100 XRP and get more USD than asked for.
+        env (offer (alice, USD (100), XRP (100)), json(jss::Flags, tfSell));
+
+        auto jrr = ledgerEntryState (env, alice, gw, "USD");
+        BEAST_EXPECT(jrr[jss::node][sfBalance.fieldName][jss::value] == "-200");
+        jrr = ledgerEntryRoot (env, alice);
+        BEAST_EXPECT(
+            jrr[jss::node][sfBalance.fieldName] == XRP (250).value ().getText ()
+        );
+
+        jrr = ledgerEntryState (env, bob, gw, "USD");
+        BEAST_EXPECT(jrr[jss::node][sfBalance.fieldName][jss::value] == "-300");
+    }
+
+    void
+    testcaseGatewayCrossCurrency()
+    {
+        testcase ("Client Issue #535: Gateway Cross Currency");
+
+        using namespace jtx;
+
+        Env env {*this};
+        auto const gw = Account {"gateway"};
+        auto const alice = Account {"alice"};
+        auto const bob = Account {"bob"};
+        auto const XTS = gw["XTS"];
+        auto const XXX = gw["XXX"];
+
+        auto const starting_xrp = XRP (100.1) +
+            env.current ()->fees ().accountReserve (1) +
+            env.current ()->fees ().base * 2;
+
+        env.fund (starting_xrp, gw, alice, bob);
+
+        env (trust (alice, XTS (1000)));
+        env (trust (alice, XXX (1000)));
+        env (trust (bob, XTS (1000)));
+        env (trust (bob, XXX (1000)));
+
+        env (pay (gw, alice, alice["XTS"] (100)));
+        env (pay (gw, alice, alice["XXX"] (100)));
+        env (pay (gw, bob, bob["XTS"] (100)));
+        env (pay (gw, bob, bob["XXX"] (100)));
+
+        env (offer (alice, XTS (100), XXX (100)));
+
+        //WS client is used here because the RPC client could not
+        //be convinced to pass the build_path argument
+        auto wsc = makeWSClient(env.app().config());
```

Comment:
Should we consider adding a `TODO` or `FIXME` or somesuch marker here? @vinniefalco?